File: app/src/main/java/betrip/kitttn/architecturecomponentstest/vm/search_results/SearchResultsViewModel.java
package betrip.kitttn.architecturecomponentstest.vm.search_results;

import android.arch.lifecycle.ViewModel;

import java.net.UnknownHostException;

import betrip.kitttn.architecturecomponentstest.services.CountryNamesLoader;
import betrip.kitttn.architecturecomponentstest.vm.search_results.state.ErrorSearchResultState;
import betrip.kitttn.architecturecomponentstest.vm.search_results.state.InitialSearchResultState;
import betrip.kitttn.architecturecomponentstest.vm.search_results.state.LoadingSearchResultState;
import betrip.kitttn.architecturecomponentstest.vm.search_results.state.SearchResultState;
import betrip.kitttn.architecturecomponentstest.vm.search_results.state.SuccessSearchResultState;
import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.BehaviorSubject;

/**
 * @author kitttn
 */

public class SearchResultsViewModel extends ViewModel {
    private CountryNamesLoader loader;
    private BehaviorSubject<SearchResultState> state = BehaviorSubject.createDefault(new InitialSearchResultState());
    private CompositeDisposable composite = new CompositeDisposable();

    public SearchResultsViewModel(CountryNamesLoader loader) {
        this.loader = loader;
        Disposable d = loader.getCountries()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(it -> state.onNext(new SuccessSearchResultState(it)),
                        Throwable::printStackTrace);

        composite.add(d);
    }

    public Observable<SearchResultState> getSearchResults() {
        return state;
    }

    public void startSearch(String query) {
        Completable completable = query.isEmpty()
                ? loader.fetchAllCountries()
                : loader.fetchCountries(query);

        state.onNext(new LoadingSearchResultState());

        Disposable d = completable
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(() -> {}, err -> state.onNext(createSearchError(err)));

        composite.add(d);
    }

    private ErrorSearchResultState createSearchError(Throwable error) {
        if (error instanceof UnknownHostException) {
            return new ErrorSearchResultState("No internet connection detected!",
                    ErrorSearchResultState.ErrorCode.NO_NETWORK);
        }

        return new ErrorSearchResultState(error.getMessage(), ErrorSearchResultState.ErrorCode.OTHER);
    }

    @Override protected void onCleared() {
        composite.clear();
        super.onCleared();
    }
}

File: app/src/main/java/betrip/kitttn/architecturecomponentstest/vm/SearchResultsViewModel.kt
package betrip.kitttn.architecturecomponentstest.vm

import android.arch.lifecycle.ViewModel
import betrip.kitttn.architecturecomponentstest.model.CountryName
import betrip.kitttn.architecturecomponentstest.plusAssign
import betrip.kitttn.architecturecomponentstest.services.CountryLoader
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.BehaviorSubject

/**
 * @author kitttn
 */

sealed class SearchResultState

class InitialSearchResultState : SearchResultState()

class SearchResultLoading : SearchResultState()

data class SearchResultSuccess(val data: List<CountryName>) : SearchResultState()

data class SearchResultError(val errorReason: String, val errorCode: Int) : SearchResultState() {
    companion object {
        const val ERROR_NO_NETWORK = 1
        const val ERROR_OTHER = 999
    }
}

class SearchResultsViewModel(private val countryLoader: CountryLoader) : ViewModel() {
    private val searchResults = BehaviorSubject.create<SearchResultState>()
    private val composite = CompositeDisposable()

    init {
        searchResults.onNext(InitialSearchResultState())

        composite += countryLoader.countries
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ searchResults.onNext(SearchResultSuccess(it)) },
                        { searchResults.onNext(SearchResultError(it.localizedMessage, SearchResultError.ERROR_OTHER)) })
    }

    fun startSearch(query: String) {
        if (query.isEmpty()) {
            fetchData { countryLoader.fetchAllCountries() }
            return
        }

        fetchData { countryLoader.fetchCountries(query) }
    }

    private fun fetchData(completable: () -> Completable) {
        searchResults.onNext(SearchResultLoading())
        composite += completable()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({}, Throwable::printStackTrace)
    }

    fun getSearchResults(): Observable<SearchResultState> = searchResults

    override fun onCleared() {
        composite.clear()
        super.onCleared()
    }
}